.github/workflows/build-all.yml

```diff
@@ -0,0 +1,114 @@
+name: Build Bowtie Image for all available versions
+
+on:
+  workflow_dispatch:
+
+env:
+  IMAGE_REGISTRY: ghcr.io/${{ github.repository_owner }}
+
+jobs:
+  versions:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+    outputs:
+      revisions: ${{ steps.revisions.outputs.value }}
+    steps:
+      - name: Collect available versions
+        id: revisions
+        run: |
+          tags=$(gh api \
+            -H "Accept: application/vnd.github+json" \
+            -H "X-GitHub-Api-Version: 2022-11-28" \
+            /repos/${{ github.repository }}/releases | jq -c '. | map(.tag_name) + ["main"]')
+          echo "value=$(echo $tags | jq -c .)" >> $GITHUB_OUTPUT
+
+  build:
+    needs: versions
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: true
+      matrix:
+        revision: ${{ fromJson(needs.versions.outputs.revisions) }}
+    permissions:
+      id-token: write
+      contents: read
+      attestations: write
+      packages: write
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ matrix.revision }}
+
+      - name: Install uv
+        uses: astral-sh/setup-uv@v5
+        with:
+          enable-cache: true
+
+      - name: Compute implementation name
+        id: impl
+        run: echo "name=$(echo '${{ github.repository }}' | awk -F '/' '{print $2}')" >> $GITHUB_OUTPUT
+
+      - name: Install qemu
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y qemu-user-static
```

Comment:
May as well ask: do you know if Kotlin supports cross-compiling natively rather than needing to use QEMU? (some languages like rust, c# and golang I know do and we already use that functionality for some of those)

Reply:
I don't know honestly. I can dig into that (I am not familiar with that topic - will learn something new :smiley: ).

But Kotlin here is used as a JVM language. If Java supports cross-compiling so does Kotlin (I think so at least)

Reply:
Actually, I think the answer would be `yes` because we compile kotlin into java classes which doesn't depend on the platform. So, we can remove this step I think

Reply:
Or not...something went wrong once I removed the qemu
https://github.com/bowtie-json-schema/kotlin-kmp-json-schema-validator/actions/runs/13722121362/job/38379776771?pr=2

> [1/2] STEP 6/8: RUN gradle --no-daemon --parallel dependencies > /dev/null
exec container process `/bin/sh`: Exec format error
Error: building at STEP "RUN gradle --no-daemon --parallel dependencies > /dev/null": while running runtime: exit status 1

I will try to debug it locally

Reply:
Okay, so it looks like the problem is not in Java/Kotlin but in Gradle which uses `sh` to execute scripts. And apparently, this is not possible without emulation which is provided by QEMU

Reply:
Huh, that's... strange. Shell scripts are text, so they shouldn't be arch-specific?? And I have to assume gradle isn't installing its *own* `sh` binary... is it????? If I can ask for one more time could you try to find someone who's maybe written a blog post on compiling cross-arch JVM images and maybe just confirm they ended up doing so via qemu? If so it's all fine to leave, but I'm just honestly curious.

Reply:
No, it doesn't not install its own shell. I was looking into something like that but had no luck. I will try doing it one more time. Something doesn't add up here...

Reply:
I have not found an article about creating multiarch image with Java app. However, I did  a few experiments and everything works just fine without any modifications with **docker** (I don't believe it installs QEMU automatically)
https://github.com/bowtie-json-schema/kotlin-kmp-json-schema-validator/actions/runs/13725912881/job/38392187019

But with podman and buildah we are getting the same error...
Will continue looking.

Reply:
CI is green without QEMU, however I am not yet sure how it happened (exactly the same configuration failed before...). Will dig deeper tomorrow

Reply:
I have figured out why the same changes did not work before)
I used a `build-all` workflow to test them - it collects tags, **checkouts them** and builds images. And this is the part I was missing - Dockerfile on the tag does not contain the changes I did :sweat_smile: And the build failed

Re changes:

The only important thing is `--platform=$BUILDPLATFORM` in the Dockerfile. It checks out an image that is compatible with build platform and compiles Kotlin/Java classes. After that, the resulting files are moved to the container with **target platform** which can use those classes